File: app/src/main/java/com/platdmit/simplecloudmanager/app/vm/factory/SingleElementViewModelFactory.java
package com.platdmit.simplecloudmanager.app.vm.factory;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

public class SingleElementViewModelFactory<Repo> extends ViewModelProvider.NewInstanceFactory {
    private Repo mRepo;
    private Class mRepoInterface;
    private long mElementId;

    public SingleElementViewModelFactory(Repo repo, Class<Repo> repoInterface, long elementId) {
        super();
        mRepo = repo;
        mRepoInterface = repoInterface;
        mElementId = elementId;
    }

    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        try {
            return (T)modelClass.getConstructor(mRepoInterface, long.class).newInstance(mRepo, mElementId);
        } catch (Exception e){
            try {
                return modelClass.newInstance();
            } catch (InstantiationException | IllegalAccessException g) {
                throw new RuntimeException("Cannot create an instance of " + modelClass, g);
            }
        }
    }
}

File: app/src/main/java/com/platdmit/simplecloudmanager/app/vm/factory/SingleElementViewModelFactory.kt
package com.platdmit.simplecloudmanager.app.vm.factory

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider.NewInstanceFactory

class SingleElementViewModelFactory<Repo>(private val repo: Repo, private val repoInterface: Class<Repo>, private val elementId: Long) : NewInstanceFactory() {
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        return try {
            modelClass.getConstructor(repoInterface, Long::class.javaPrimitiveType).newInstance(repo, elementId)
        } catch (e: Exception) {
            try {
                modelClass.newInstance()
            } catch (g: InstantiationException) {
                throw RuntimeException("Cannot create an instance of $modelClass", g)
            } catch (g: IllegalAccessException) {
                throw RuntimeException("Cannot create an instance of $modelClass", g)
            }
        }
    }
}